config/settings.d/tftp.yml.example

```diff
@@ -13,3 +13,6 @@
 # Defines the default certificate action for certificate checking.
 # When false, the argument --no-check-certificate will be used.
 #:verify_server_cert: true
+
+# Directory for OS specific NBPs for "Grub2 UEFI" PXE loaders
+#:bootloader_universe:
```

Comment:
Where does "universe" come from?

Reply:
The term "universe" was a design decision from our side. The meaning is: This is the place where all custom bootloader files reside.

Reply:
What would be the default value?

Or, can it be just a boolean, and the value of the path would always be a sub-directory under the `:tftproot:`?
(hard coded)

Reply:
In contrast to [my comment below](https://github.com/theforeman/smart-proxy/pull/877#discussion_r1652024829), we could set a commented out default value here like `#:bootloader_universe: /usr/local/share/bootloader-universe`. A commented out value won't affect the logic. But on the other hand I don't know if there is any benefit of adding a commented out value here as a user should always set the value with `foreman-installer --foreman-proxy-tftp-bootloader-universe` and never modify the config files directly.

Regarding the boolean value: For readability we could also add a boolean switch to explicitly enable the usage of the Bootloader Universe. But again a user should never edit the config file directly, so I would vote for one setting to be enough for the Bootloader Universe.

Note: The Bootloader Universe does not necessarily have to be a sub-directory of the `:tftproot:`. It is an absolute path that can be anywhere on your Smart Proxy.

Reply:
> Regarding the boolean value: For readability we could also add a boolean switch to explicitly enable the usage of the Bootloader Universe. 

I am a big fan of this. Explicitly enabling something and setting its value seems more natural and user-friendly to me. 

> But again a user should never edit the config file directly, so I would vote for one setting to be enough for the Bootloader Universe.

If we'll have one setting for enabling the feature and a second one for the path, that won't be a much of problem, right?

> Note: The Bootloader Universe does not necessarily have to be a sub-directory of the :tftproot:. It is an absolute path that can be anywhere on your Smart Proxy.

Yes, that's correct. We just need to think about the SElinux policies, and AFAIK, the `:tftproot:` is already part of Foreman's policies.

Reply:
> Regarding the boolean value: For readability we could also add a boolean switch to explicitly enable the usage of the Bootloader Universe.

> I am a big fan of this. Explicitly enabling something and setting its value seems more natural and user-friendly to me.

Disregard this, I had discussion with others and let's have only one setting as it is now.

Reply:
> Yes, that's correct. We just need to think about the SElinux policies, and AFAIK, the :tftproot: is already part of Foreman's policies.

I am just thinking about if there are any disadvantages of storing the Bootloader Universe inside the `:tftproot:`. I Maybe we should even suggest this in the documentation as the preferred way. :thinking:

Reply:
I certainly think we need to come up with a recommended way of using it.

To me universe is a bit confusing because it immediately makes the think of Ubuntu Universe repos, but that may just be me.

Now the question is: who provides these? Is it done via packages, manually from the user, automated somehow? Mixed?

As I read it, the content is essentially data files. They may be binaries and for the target system architecture dependent, but for the host they're independent: they'll just be served to hosts. Is that correct?

If we look at the [Linux File Hierachy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html) it describes what directories are used for. Then we come to:

* Packages: `/usr/share`
* User provided: `/usr/local/share` or arguably `/var/lib`
* Automatically managed: `/var/lib`
* Mixed: you'll need multiple load paths to combine them

Small performance factor: copying files on the same filesystem can be more efficient on filesystems that implement Copy-on-Write, but Ruby doesn't appear to support reflinks out of the box. Given how small the files are, I think we should ignore that for now. I saw some copying of duplicate files which could at least use hardlinks, but we may be micro-optimizing there.

On the installer side: I'd urge you to find a default value for it so it just works. If it's opt-in, preferably make it a single boolean that ensures it works. Opt-out may be preferred so there's fewer things a user needs to set up.

Reply:
> To me universe is a bit confusing because it immediately makes the think of Ubuntu Universe repos, but that may just be me.

What about naming it bootloader_binaries or OS_boot_binaries?

> Small performance factor: copying files on the same filesystem can be more efficient on filesystems that implement Copy-on-Write, but Ruby doesn't appear to support reflinks out of the box...

We proposed using relative symlinks instead of copying these files to address that issue.

> On the installer side: I'd urge you to find a default value for it so it just works. If it's opt-in, preferably make it a single boolean that ensures it works. Opt-out may be preferred so there's fewer things a user needs to set up.

We decided to place this directory by default under `:tftpboot/bootloader_universe/pxegrub2/` and to use it as the default. If a user does not deploy any bootloader files in this directory, the default files from `:tftproot:/grub2/` will be linked to `:tftproot:/host_config/<MAC>/grub2`.

Reply:
> We decided to place this directory by default under `:tftpboot/bootloader_universe/pxegrub2/` and to use it as the default. If a user does not deploy any bootloader files in this directory, the default files from `:tftproot:/grub2/` will be linked to `:tftproot:/host_config/<MAC>/grub2`.

If it's always hosted under the `tftproot` then why do we need an additional setting? Is there a use case where it helps?

Reply:
We don't need the additional setting. We are planning on removing it.